### PR TITLE
feature(alternator): Add a performance test

### PR DESCRIPTION
**Pull request by fruch**
This PR introduce performance tests for alternator
* [x] throughput: write, read, mix 
* [x] latency
* [x] option to enable/disable LWT to compare results
* [x] refactor email to make it more clear what is the scenario 

until implementing the DNS solution, each loader select a host to work with in a round-robin

### CODE REVIEW

**Comment by fruch** [comment]
@roydahan there a reason why this is still in Draft

also the data is going into it's own index, since we can't add anything to old index mapping

**Comment by fruch** [comment]
new jobs:
https://jenkins.scylladb.com/job/scylla-master/job/performance-tests/

**Comment by bentsi** [comment]
add timestamp to the subject of the email

File: performance_regression_alternator_test.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            self.create_test_stats(test_name=test_name, sub_type=sub_type,
+                                   doc_id_with_timestamp=True, append_sub_test_to_name=False)
+        stress_queue = self.run_stress_thread(stress_cmd=stress_cmd, stress_num=stress_num, keyspace_num=keyspace_num,
+                                              prefix=prefix, stats_aggregate_cmds=False)
+        self.get_stress_results(queue=stress_queue, store_results=True)
+        if save_stats:
+            self.update_test_details(scylla_conf=True)
+            # self.display_results(results, test_name=test_name)
+            if check_regression:
+                self.check_regression()
```
Comment (aleksbykov) [inline performance_regression_alternator_test.py]: May remove this code, and appropriate parameter from signature? because it is not used in code

**Comment by bentsi** [comment]
please, fix conflicts

File: vars/perfRegressionParallelPipeline.groovy [hunk 2]
```diff
@@ -64,13 +64,15 @@ def call(Map pipelineParams) {
                         } else {
                             sub_tests = [pipelineParams.test_name]
                         }
-                        for (sub_test in sub_tests) {
+                        for (t in sub_tests) {
                             def perf_test
+                            def sub_test = t
```
Comment (bentsi) [inline vars/perfRegressionParallelPipeline.groovy]: why this trick is necessary?

File: sdcm/results_base.html [hunk 3]
```diff
@@ -140,7 +140,40 @@
...
+        {% block extra_link %}
+        {% endblock %}
+        {% if grafana_screenshots %}
+            {% if grafana_screenshots[0] %}
+                <li><a href={{ grafana_screenshots[0] }}>Download "Per server metrics nemesis" Grafana Screenshot</a></li>
+            {% endif %}
+            {% if grafana_screenshots[1] %}
+                <li><a href={{ grafana_screenshots[1] }}>Download "Overview metrics" Grafana Screenshot</a></li>
+            {% endif %}
+            {% if grafana_screenshots[2] %}
```
Comment (bentsi) [inline sdcm/results_base.html]: please check if we didn't fail with Index out of range error

File: sdcm/results_analyze.py [hunk 4]
```diff
@@ -669,13 +697,16 @@ def check_regression(self, test_id, is_gce=False):
...
+                       ycsb_raw_cmd=ycsb.get('raw_cmd', "") if ycsb else "",
                        job_url=doc['_source']['test_details'].get('job_url', ""),
                        dashboard_master=self.gen_kibana_dashboard_url(dashboard_path),
                        )
         self.log.debug('Regression analysis:')
         self.log.debug(PP.pformat(results))
         test_name = full_test_name.split('.')[-1]  # Example: longevity_test.py:LongevityTest.test_custom_time
         subject = 'Performance Regression Compare Results - {} - {}'.format(test_name, test_version)
+        if ycsb:
+            subject = '(Alternator) Performance Regression - {} - {}'.format(test_name, test_version)
```
Comment (fruch) [inline sdcm/results_analyze.py]: add the data like in `check_regression_baseline`

File: performance_regression_alternator_test.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            return total_ops, latency_write_99_avg, latency_write_99_stdev
+        return None
+
+    @property
+    def alternator_endpoint_url(self):
+        alternator_port = self.params.get('alternator_port')
+        endpoint_url = f'http://{normalize_ipv6_url(self.db_cluster.nodes[0].external_address)}:{alternator_port}'
+        return endpoint_url
+
+    def crate_alternator_table(self, table_params):
```
[the remark this answers is not in the document]
Comment (fruch) [reply performance_regression_alternator_test.py]: fixed

File: performance_regression_alternator_test.py (hunk 1 above)
Comment (fruch) [reply performance_regression_alternator_test.py]: fixed

File: vars/perfRegressionParallelPipeline.groovy (hunk 2 above)
Comment (fruch) [reply vars/perfRegressionParallelPipeline.groovy]: cause of parallel jobs in the jenkinsfile

File: sdcm/results_base.html (hunk 3 above)
Comment (fruch) [reply sdcm/results_base.html]: I've check, and this would return False, if IndexError is raised

File: sdcm/results_analyze.py (hunk 4 above)
Comment (fruch) [reply sdcm/results_analyze.py]: done

**Comment by fruch** [comment]
added jobs:
* https://jenkins.scylladb.com/view/master/job/scylla-master/job/scylla-master-perf-regression-alternator-latency-test/
* https://jenkins.scylladb.com/view/master/job/scylla-master/job/scylla-master-perf-regression-alternator-throughput-test/
* https://jenkins.scylladb.com/view/Performance/job/Perf-Regression/job/scylla-release-perf-regression-alternator-latency-test/
* https://jenkins.scylladb.com/view/Performance/job/Perf-Regression/job/scylla-release-perf-regression-alternator-throughput-test/

**Comment by fruch** [comment]
@roydahan @bentsi, also we need to decide on the branching for pref tests, how we are going to handle them.

File: test-cases/performance/perf-regression-alternator.100threads.30M-keys.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+prepare_write_cmd:
+  - >-
+    bin/ycsb load dynamodb -P workloads/workloada -threads 150 -p recordcount=30000000
+    -p requestdistribution=uniform -p fieldcount=10 -p fieldlength=256
+    -p insertorder=hashed -p insertstart=0 -p insertcount=10000000
+
+  - >-
+    bin/ycsb load dynamodb -P workloads/workloada -threads 150 -p recordcount=30000000
+    -p requestdistribution=uniform -p fieldcount=10 -p fieldlength=256
+    -p insertorder=hashed -p insertstart=10000000 -p insertcount=10000000
```
Comment (roydahan) [inline test-cases/performance/perf-regression-alternator.100threads.30M-keys.yaml:13]: shouldn't it be insertstart=10000001?
(also in all other places)

File: performance_regression_test.py [hunk 7]
```diff
@@ -433,17 +436,13 @@ def test_mixed(self):
...
         # create new document in ES with doc_id = test_id + timestamp
         # allow to correctly save results for future compare
         self.create_test_stats(sub_type='write-prepare', doc_id_with_timestamp=True)
         self.run_fstrim_on_all_db_nodes()
         # run a write workload as a preparation
-        stress_queue = self.run_stress_thread(stress_cmd=base_cmd_w, stress_num=2, prefix='preload-',
-                                              stats_aggregate_cmds=False)
-        self.get_stress_results(queue=stress_queue, store_results=False)
-        self.update_test_details()
+        self.preload_data()
```
[the remark this answers is not in the document]
Comment (fruch) [reply performance_regression_test.py:445]: reverted

File: test-cases/performance/perf-regression-alternator.100threads.30M-keys.yaml (hunk 6 above)
Comment (fruch) [reply test-cases/performance/perf-regression-alternator.100threads.30M-keys.yaml:13]: this is how it should be setup see:
https://github.com/brianfrankcooper/YCSB/wiki/Running-a-Workload-in-Parallel

File: performance_regression_alternator_test.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+            self.update_test_details()
+        else:
+            self.log.warning("No prepare command defined in YAML!")
+
+    def test_write(self):
+        """
+        Test steps:
+
+        1. Run a write workload without lwt
```
[the remark this answers is not in the document]
Comment (fruch) [reply performance_regression_alternator_test.py]: fixed

File: performance_regression_alternator_test.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        self._workload(sub_type='with-lwt', stress_cmd=base_cmd_w, stress_num=stress_multiplier, keyspace_num=1)
+
+        self.check_regression_with_baseline('cql')
+
+    def test_read(self):
+        """
+        Test steps:
+
+        1. Run a write workload as a preparation
+        2. Run a read workload without lwt
```
[the remark this answers is not in the document]
Comment (fruch) [reply performance_regression_alternator_test.py]: fixed

File: performance_regression_alternator_test.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+        self._workload(sub_type='with-lwt', stress_cmd=base_cmd_r, stress_num=stress_multiplier, keyspace_num=1)
+
+        self.check_regression_with_baseline('cql')
+
+    def test_mixed(self):
+        """
+        Test steps:
+
+        1. Run a write workload as a preparation
+        2. Run a mixed workload without lwt
```
[the remark this answers is not in the document]
Comment (fruch) [reply performance_regression_alternator_test.py]: fixed

**Comment by fruch** [comment]
since this is going to it's own branch, no going to backport this to branch-4.0 (maybe only part of it, relevant to any alternator change that might be needed)

**Pull request merged**